File: typescript/src/astro/naming-region-origins.ts
/**
 * Origins of Elite Dangerous **naming regions** and the fallback that synthesises
 * an origin for any procedural sector.
 *
 * Encoding a system name to an `id64` needs its region's origin (in internal
 * units, 32 per light-year, measured from the galaxy corner). Hand-authored
 * regions — nebula/cluster sectors such as `Col 285 Sector` or
 * `Cepheus Dark Region B` — have catalogued origins that are *not* aligned to the
 * boxel grid, so they are looked up from data. Any other region is a plain
 * procedural sector whose origin is simply its grid position times the sector
 * size.
 *
 * The catalogue derives from the EDTS reference implementation and community procedural-naming
 * research, cross-checked against EDSM/Spansh; see [`data/astro/SOURCES.md`](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/data/astro/SOURCES.md) for
 * provenance.
 *
 * The catalogue is loaded once from shared JSON ([`data/astro/`](https://github.com/DarkSession/Elite-Dangerous-Almanac/tree/main/data/astro)); the lookup map is
 * an immutable module constant.
 *
 * @packageDocumentation
 */

import { sectorGridPositionFromName, sectorNameFromGridPosition } from './sector-name.js';
import { SECTOR_INTERNAL_SIZE } from './system-address.js';
import originsData from '../../../data/astro/named-region-origins.jsonc' with { type: 'json' };
import { deepFreeze } from '../internal/deep-freeze.js';
import { normalizeKey } from '../internal/registry-index.js';

export { SECTOR_INTERNAL_SIZE } from './system-address.js';

/**
 * A region's origin and extent, in internal units (32 per light-year, measured
 * from the galaxy corner).
 *
 * @example
 * ```ts
 * import type { NamingRegionOrigin } from '@elite-dangerous-almanac/core/astro/naming-region-origins';
 *
 * const origin: NamingRegionOrigin = {
 *   name: 'Example Sector',
 *   x0: 0, y0: 0, z0: 0,
 *   sizeX: 40960, sizeY: 40960, sizeZ: 40960,
 * };
 * ```
 */
export interface NamingRegionOrigin {
    /** The region name, as catalogued. */
    readonly name: string;
    /** Origin X in internal units. */
    readonly x0: number;
    /** Origin Y in internal units. */
    readonly y0: number;
    /** Origin Z in internal units. */
    readonly z0: number;
    /** Extent along X in internal units. */
    readonly sizeX: number;
    /** Extent along Y in internal units. */
    readonly sizeY: number;
    /** Extent along Z in internal units. */
    readonly sizeZ: number;
}

const CATALOGUE: ReadonlyMap<string, NamingRegionOrigin> = new Map(
    deepFreeze(originsData as readonly NamingRegionOrigin[]).map((r) => [r.name.toLowerCase(), r]),
);

/**
 * Look up a hand-authored named region's catalogued origin.
 *
 * @param name - A named region in any casing, with optional surrounding whitespace.
 * @returns Its canonical origin record, or `null` when it is not catalogued.
 * @example
 * ```ts
 * import { getHandAuthoredRegionOrigin } from '@elite-dangerous-almanac/core/astro/naming-region-origins';
 *
 * getHandAuthoredRegionOrigin('  PLEIADES SECTOR ')?.name; // -> 'Pleiades Sector'
 * getHandAuthoredRegionOrigin('Synuefe');                  // -> null
 * ```
 */
export function getHandAuthoredRegionOrigin(name: string): NamingRegionOrigin | null {
    return CATALOGUE.get(normalizeKey(name)) ?? null;
}

/**
 * Resolve a region name to its origin.
 *
 * Hand-authored regions come from the catalogue; any other name is treated as a
 * procedural sector and its origin synthesised from the sector grid. Returns
 * `null` only when the name is neither catalogued nor a valid procedural sector.
 *
 * @param name - A region (sector) name in any casing.
 * @returns The region origin, or `null` if the name cannot be resolved.
 * @example
 * ```ts
 * import { resolveNamingRegionOrigin } from '@elite-dangerous-almanac/core/astro/naming-region-origins';
 *
 * resolveNamingRegionOrigin('Pleiades Sector'); // catalogued, hand-authored origin
 * resolveNamingRegionOrigin('Synuefe');         // origin derived from its grid position
 * resolveNamingRegionOrigin('not a region');    // -> null
 * ```
 */
export function resolveNamingRegionOrigin(name: string): NamingRegionOrigin | null {
    const namedOrigin = getHandAuthoredRegionOrigin(name);
    if (namedOrigin) return namedOrigin;

    const position = sectorGridPositionFromName(name);
    if (!position) return null;

    return {
        name: sectorNameFromGridPosition(position),
        x0: position.sectorX * SECTOR_INTERNAL_SIZE,
        y0: position.sectorY * SECTOR_INTERNAL_SIZE,
        z0: position.sectorZ * SECTOR_INTERNAL_SIZE,
        sizeX: SECTOR_INTERNAL_SIZE,
        sizeY: SECTOR_INTERNAL_SIZE,
        sizeZ: SECTOR_INTERNAL_SIZE,
    };
}

File: dotnet/src/EliteDangerousAlmanac/Astronomy/NamingRegionOrigins.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;
using EliteDangerousAlmanac.Internal;

namespace EliteDangerousAlmanac.Astronomy;

/// <summary>Where one naming region starts, and how far it reaches.</summary>
/// <param name="Name">The region name, as catalogued.</param>
/// <param name="X">The starting X coordinate in internal units.</param>
/// <param name="Y">The starting Y coordinate in internal units.</param>
/// <param name="Z">The starting Z coordinate in internal units.</param>
/// <param name="SizeX">How far the region reaches along X, in internal units.</param>
/// <param name="SizeY">How far the region reaches along Y, in internal units.</param>
/// <param name="SizeZ">How far the region reaches along Z, in internal units.</param>
/// <remarks>
/// Internal units are thirty-seconds of a light year, counted from the galaxy corner.
/// </remarks>
public sealed record NamingRegionOrigin(
    string Name,
    [property: JsonPropertyName("x0")] double X,
    [property: JsonPropertyName("y0")] double Y,
    [property: JsonPropertyName("z0")] double Z,
    double SizeX,
    double SizeY,
    double SizeZ);

/// <summary>Where the game's naming regions start.</summary>
/// <remarks>
/// <para>
/// Writing a system name as an address needs the region's own starting point. A region the
/// game names by hand, such as <c>Col 285 Sector</c> or <c>Cepheus Dark Region B</c>,
/// starts somewhere the boxel grid does not, so its starting point is catalogued. Every
/// other region is a plain procedural sector, whose starting point is its place on the grid
/// times the sector size.
/// </para>
/// <para>
/// The catalogue loads from its shared data file on first use, and every record is
/// immutable. It comes from the EDTS reference and the community research behind it. See
/// <c>data/astro/SOURCES.md</c> for where it was read from, and <c>ATTRIBUTIONS.md</c> for
/// credit and licence terms.
/// </para>
/// </remarks>
public static class NamingRegionOrigins
{
    private static readonly Lazy<IReadOnlyList<NamingRegionOrigin>> Catalogue = new(
        () => SharedData.LoadList<NamingRegionOrigin>("data/astro/named-region-origins.jsonc"));

    private static readonly Lazy<IReadOnlyDictionary<string, NamingRegionOrigin>> ByName = new(
        () => RegistryIndex.CreateKeyIndex(HandAuthored, region => region.Name));

    /// <summary>Every region the game names by hand.</summary>
    public static IReadOnlyList<NamingRegionOrigin> HandAuthored => Catalogue.Value;

    /// <summary>Finds where one region the game names by hand starts.</summary>
    /// <param name="name">The region name, in any casing, with any surrounding whitespace.</param>
    /// <returns>
    /// The catalogued record, or <see langword="null"/> where no such region is catalogued.
    /// An absent name is a miss, answered the way an unknown one is.
    /// </returns>
    public static NamingRegionOrigin? FindHandAuthored(string? name) =>
        RegistryIndex.FindInKeyIndex(ByName.Value, name);

    /// <summary>Reads where one region starts, whether the game names it or the grid does.</summary>
    /// <param name="name">The region name, in any casing, with any surrounding whitespace.</param>
    /// <returns>
    /// The region's starting point, or <see langword="null"/> where the name is neither
    /// catalogued nor a procedural sector.
    /// </returns>
    public static NamingRegionOrigin? Resolve(string? name)
    {
        NamingRegionOrigin? handAuthored = FindHandAuthored(name);
        if (handAuthored is not null) return handAuthored;

        SectorGridPosition? position = SectorName.ToGridPosition(name);
        if (position is null) return null;

        return new NamingRegionOrigin(
            SectorName.FromGridPosition(position),
            (double)position.SectorX * SystemAddress.SectorInternalSize,
            (double)position.SectorY * SystemAddress.SectorInternalSize,
            (double)position.SectorZ * SystemAddress.SectorInternalSize,
            SystemAddress.SectorInternalSize,
            SystemAddress.SectorInternalSize,
            SystemAddress.SectorInternalSize);
    }
}
